Motion controller examples.

taylorjames9 opened the pull request:
I've added a folder to the Dev2017 branch that contains easy to use components for grab/scale/rotate/throw. Folder is located in: Assets\HoloToolkit\Input\Tests\CK_GrabMechanics  Test scene located in Assets\HoloToolkit\Input\Tests\CK_GrabMechanics\Scenes ***Right now all of this is HUX/MRTK independent, so it does not rely on Paseb's merge.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/MixedRealityToolkit-Unity?pullRequest=1021) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/MixedRealityToolkit-Unity?pullRequest=1021) before we can accept your contribution.<br/><br/>:x: taylorjames9 [sign now](https://cla.opensource.microsoft.com/Microsoft/MixedRealityToolkit-Unity?pullRequest=1021)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/MixedRealityToolkit-Unity?pullRequest=1021) it.</sub>

StephenHodgson commented on Packages/manifest.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Please remove this and it's meta file from the PR

StephenHodgson commented on Assets/HoloToolkit/Input/Tests/CK_GrabMechanics.meta:
```diff
@@ -0,0 +1,10 @@
+fileFormatVersion: 2
```
These files should go in `Assets/HoloToolkit-Examples/`

taylorjames9 commented:
I think that should do it. I made the changes that you requested. I've also linked my microsoft account with my github account to take care of the CLA. Let me know if you have everything that you need. Again, this pull is independent of the MRTK/HUX merge that Paseb submitted.

StephenHodgson commented:
CK_GrabMechanics?

taylorjames9 commented:
I noticed that underscores aren't used in the project so I renamed that folder to CKGrabMechanics and put it in the HoloToolkit-Examples folder.

StephenHodgson commented:
That does CK stand for?

taylorjames9 commented:
actually i'm updating the name one more time from CK_GrabMechanics to WindowsMRControllers-GrabMechanics

jessemcculloch commented:
Most likely Crystal Key, the name of the motion controllers

StephenHodgson commented:
Agreed, I think a name change is in order. Just for clarification.

taylorjames9 commented:
ok - should be good to go as of the latest commit 21c7798. Let me know if there's anything else I can do.

cre8ivepark commented:
Thanks James!
I have added MRTK logo and will do some visual fit & finish.

StephenHodgson commented on Assets/HoloToolkit-Examples/WindowsMRControllers-GrabMechanics/Textures/MRTK_Logo.png.meta:
```diff
@@ -1,18 +1,20 @@
 fileFormatVersion: 2
-guid: e7c1f473df573ff4094e55c183149d96
```
Do we want this in the grab mechanics examples folder, or actually have it in it's own folder in Examples with other toolkit branding assets?

taylorjames9 commented:
Nice, Yoon! Here are some of the updates I will provide as soon as they are stable: 

-Should be able to enter an object collider & then grab (right now you have to grab _then_ enter)
-Switching thumbstick rotation example to a touchpad rotation (bc thumbstick is reserved for teleport)
-If it's easy to "drop in" I can also add some simple teleport behaviour to the scene to allow user to grab things that might be otherwise out of reach
-by default, the user should not be able to grab multiple objects at a time
-Changing the name of the folder from WindowsMotionController-GrabMechanics to MotionControllers-GrabMechanics

cre8ivepark commented:
I have updated the scene with some visual fit & finish including text labels. 
![r2017-09-28 12_51_37-mixed reality portal](https://user-images.githubusercontent.com/13754172/30988990-ff6061f6-a450-11e7-9b2d-8e988269e327.png)
![r2017-09-28 12_51_44-mixed reality portal](https://user-images.githubusercontent.com/13754172/30988989-ff604324-a450-11e7-990a-c288931cb780.png)
![r2017-09-28 12_52_02-mixed reality portal](https://user-images.githubusercontent.com/13754172/30988991-ff6183ce-a450-11e7-809b-443758387f99.png)

taylorjames9 commented:
After merging code on my fork from a couple of different contributors today, I'm finding the examples are in an awkward state. My hope is that in a day or two, when we get the project in a nice stable state, we can ping you to take a fresh look at the pull request. Hope that's alright. Best Regards, -James

taylorjames9 commented:
@cre8ivepark the new look is such an improvement!

StephenHodgson commented:
@taylorjames9, no worries, I'm excited to see the final!  I've got some motion controllers to help test this out too.

DDReaper commented:
I take it this also works with gamepad input and not just MR Controllers?
Will try this out later this weekend

StephenHodgson commented:
@taylorjames9 it seems you've been committing directly to the Dev branch instead of this PR for some of the changes you've requested.  Please be sure to double check which repository/branch you're committing to, before pushing your changes. 

Thanks.

StephenHodgson commented on Assets/HoloToolkit-Examples/MotionControllers-GrabMechanics/Scripts/ScalableTwoHand.cs:
```diff
@@ -0,0 +1,16 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class ScalableTwoHand : MonoBehaviour {
```
Why is there an empty class?

taylorjames9 replied:
I just realized my comment above applies to something different than what you're referring to. That's my mistake. So in my last comment, I was talking about the two-hand rotatable object that relies almost exclusively on Unity joints (an example that we've since disabled in the scene because it's very finicky). I see now that you're referring to ScalableTwoHand.cs, which is a script that has been replaced by the combination of GrabbableMultiJoint.cs and ScalableObject.cs. I've deleted ScalableTwoHand.cs.

taylorjames9 commented:
Really good question. So there's an empty class because 1.) we want user's to get in the habit of wanting a behaviour, and then attaching a script component with a similar name. 2.) right now we are using UNity's joints for the two hand grab -- so there's nothing in the script. This is temporarily useful, but ultimately we want to transition away from writing Unity joints, and write our own. (Unity Joints are kinda finicky and can go haywire we have found.) When we write our own, the empty script will host some code for this new type of joint. I can delete the empty class in the meantime if it's too confusing for users.

StephenHodgson commented on Assets/HoloToolkit-Examples/MotionControllers-GrabMechanics/Scripts/RotatableObject.cs:
```diff
@@ -17,28 +17,35 @@ public class RotatableObject : BaseUsable
         /// <param name="obj"></param>
+        /// 
+
+        protected void Start()
```
should this be virtual?

taylorjames9 replied:
I didn't see a Start() method in the latest version of that script, but I did see a similar, non-virtual method for Awake() with no content. I've deleted the empty function call.

StephenHodgson commented on Assets/HoloToolkit-Examples/MotionControllers-GrabMechanics/Textures/MRTK_Logo.png.meta:
```diff
@@ -0,0 +1,94 @@
+fileFormatVersion: 2
+guid: f721996453d888a4db83f0f9f1a4eb7c
```
We should move this into it's own folder most likely, seeing as it's for the toolit in general.

taylorjames9 replied:
Sure - what folder should it be in?

taylorjames9 replied:
I've moved it to HoloToolKit/UI/Textures/

StephenHodgson replied:
That's exactly what I was thinking! :)

cre8ivepark commented:
Scene Video - https://gfycat.com/gifs/detail/EnlightenedDefinitiveGelding

taylorjames9 commented:
Ok. This is now stable enough to pull-in to Dev_Unity2017.2 Several known issues that are out of our control at the moment include: 1.) on the touch rotate example, the Rotatable Object seems to only respond to the left touchpad, despite checks for "handedness" associated with the grabbing controller 2.) the forward of the controllerCubes (user's hands) sometimes gets offset from the true forward of the controllers. Thanks for your help. Let me know if there are any other fixes/adjustments I can provide.

StephenHodgson commented on ProjectSettings/EditorBuildSettings.asset:
```diff
@@ -4,4 +4,7 @@
-  m_Scenes: []
+  m_Scenes:
+  - enabled: 1
+    path: Assets/HoloToolkit-Examples/MotionControllers-GrabMechanics/Scenes/GrabMechanics.unity
```
Nit: revert

taylorjames9 replied:
Do you want me to remove that scene from BuildSettings and recommit?

StephenHodgson replied:
Sorry about the late reply, but yes.

taylorjames9 replied:
Ok reverted EditorBuildSettings so there are 0 scenes and committed. Let me know if there's anything else.

StephenHodgson commented on Assets/HoloToolkit/UI/Textures/MRTK_Logo.png.meta:
```diff
@@ -0,0 +1,94 @@
+fileFormatVersion: 2
+guid: f721996453d888a4db83f0f9f1a4eb7c
```
Images should be a power of 2, so mipmapping works better (eg, 64, 128, 512, 1024, 2048, etc)

cre8ivepark replied:
@StephenHodgson Updated the image file to 2048x512px. Thanks!

DDReaper commented:
Darn that project settings file.  It's probably good to merge regardless and fix it again later.  Or simply revert changes to that file and let GIT sort it out.

StephenHodgson commented on ProjectSettings/ProjectVersion.txt:
```diff
@@ -1 +1 @@
-m_EditorVersion: 2017.2.0f2
+m_EditorVersion: 2017.2.0f1
```
![image](https://user-images.githubusercontent.com/13334553/31206885-80fed638-a92f-11e7-9739-8a4f6f086ead.png)

StephenHodgson replied:
Easier just to revert all changes

taylorjames9 commented:
Let me know if there is anything hindering this pull request from going through. I believe @cre8ivepark is adding one more diagram to the Readme, and I'm going to do some more testing with different combinations of scripts. Both of those additions are polish. I think it's pretty much good to go unless I hear differently.

StephenHodgson commented:
I haven't had a chance to review it yet.

taylorjames9 commented:
No worries (and no pressure). I just wanted to make sure I wasn't missing something you need. Have a great weekend.

DDReaper commented:
Is this ready to test? I have my controllers now, so I'm eager to take them for a spin :D

taylorjames9 commented:
@DDReaper Yup. Should be in good shape to test out. Find the scene here: Assets/HoloToolkit-Examples/MotionController-GrabMechanics/Scenes/GrabMechanics.unity

DDReaper commented:
Actually side note. Why are the tables positioned above the floor? why not just size them for waist height?

StephenHodgson commented on UnityPackageManager/Project/manifest.json:
```diff
@@ -0,0 +1,4 @@
+{
+	"dependencies": {
```
Please remove this file and its meta

StephenHodgson commented:
@taylorjames9 did the 2017.2.0f3 MRTP fix issue 1 for you?

https://forum.unity.com/threads/custom-build-2017-2-rc-mrtp-windows-mixed-reality-technical-preview.498253/

> Fix for buttons not working with InputManager when only the left controller is on

taylorjames9 commented:
@StephenHodgson I don't think that's the same issue, but let me try an upstream pull and see if that fixes it.

DDReaper commented:
Will test with f2 and then upgrade to f3 to see if it helps. (will check the controller code to ensure it's not limited there)

taylorjames9 commented:
@DDReaper I think that "Temp" might be a reserved folder name in Unity. Anyway, I've deleted the Temp folder altogether. I've also gotten rid of unused variables in BaseScalable.cs. I have deleted an assigned by unused variable on GrabbableColor.cs. As for the images in the ReadMe, can you see them if you go here? [](https://github.com/taylorjames9/MixedRealityToolkit-Unity/tree/Dev_Unity_2017.2.0/Assets/HoloToolkit-Examples/MotionControllers-GrabMechanics)

The ReadMe is specific for this specific examples folder. i"ve never had any trouble viewing them before. I'm going to take a moment to do another deep dive into the left touchpad issue.

taylorjames9 commented:
Touchpad rotation now works with the correct hand. Are we still trying to stay on Unity2017.2.0f1?

keveleigh commented:
@taylorjames9, nope, #1143 is updating to the latest public 2017.2 MRTP, so feel free to take that change!

DDReaper commented:
What's stopping this being merged now @StephenHodgson ?
If it's the project version, it may be easier just to fix that after merge, rather than rebase and potentially corrupt the PR.  I've tested this a lot over the weekend without issue on the current commit of the dev branch.

StephenHodgson commented:
I just haven't had time to review it is all. 

Maybe we should update the target branch to be master now that we're merging dev into master?

StephenHodgson commented:
@jamesashley1 testing this meow

StephenHodgson commented on UnityPackageManager/Project/manifest.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Please remove the `Project/Manifest` files
